### ISSUE TITLE
Question: is Zenoh intended to work over rapidly churning networks?

### ISSUE THREAD

**Issue by ckaran** [eclipse-zenoh/zenoh#113]
I ran across Zenoh recently, and think it may be useful for a project I'm working on.  My issue is that data needs to be replicated across a network that is constantly experiencing severe churn (all nodes are highly mobile and only intermittently in contact with one another).  Network partitions are the norm, and can last from seconds to hours, with little or no warning.  Moreover, routes are changing moment by moment; the only thing that I *think* might work would be multicast UDP with a custom gossip protocol, but I'm not sure yet about that part.  The real question is, can Zenoh as a whole handle this kind of constant churn?  Or is it really intended for more stable networks?

**Comment by OlivierHecart** [comment]
Zenoh clearly targets dynamic networks where nodes are constantly joining, leaving and or moving and where the network topology changes frequently.

From our past experience UDP multicast is often poorly supported especially by radio based networks (even WiFi) and by mobile devices. It is not suitable to reliably transport data payload. That’s why, zenoh leverages UDP multicast for dynamic discovery only in combination with a gossip discovery. Data goes through unicast communications (TCP, UDP unicast, etc …)

zenoh supports a variety of communication models including peer-to-peer as well as routed communications. A LinkState protocol is used to discover and monitor the network topology and compute data routes. zenoh constantly monitor network topology changes and adapts to them (joining/leaving nodes, joining/leaving sub-systems, broken/new links, etc …)

We ran some experiments with a 100 nodes network (low latency). It takes approximatively 1 second for all nodes in the network to discover a topology change and recompute data routes. During this second, zenoh continues to operate, but some subscribers in the system may miss some data samples. An end-to-end reliability can be set up to overcome this problem.

**Comment by ckaran** [comment]
> UDP multicast is often poorly supported especially by radio based networks (even WiFi) and by mobile devices.

My apologies, I meant to say UDP **broadcast.**  That doesn't change your later comments, just fixing my mistake...

> From our past experience UDP multicast is often poorly supported especially by radio based networks (even WiFi) and by mobile devices. It is not suitable to reliably transport data payload. That’s why, zenoh leverages UDP multicast for dynamic discovery only in combination with a gossip discovery. Data goes through unicast communications (TCP, UDP unicast, etc …)

There may be a better way than this.  I used [raptorq](https://crates.io/crates/raptorq) in my dissertation work to packetize large messages.  This increased the number of bits broadcast over the air, but also significantly improved reliability as the loss of a few packets didn't mean I lost the entire message (all my work was done over unreliable, UDP broadcast-like simulated networks).  What I didn't have a chance to do was to combine this with [Triangular Network Coding (TNC)](https://en.wikipedia.org/wiki/Triangular_network_coding).  In theory, if you use TNC to combine several messages together, and then encode & packetize them with raptorq, both your latency and throughput will improve.  Note that I have not yet had a chance to test this out, so test, test, test before committing to it!

> Zenoh clearly targets dynamic networks where nodes are constantly joining, leaving and or moving and where the network topology changes frequently.  
> ...
> We ran some experiments with a 100 nodes network (low latency). It takes approximatively 1 second for all nodes in the network to discover a topology change and recompute data routes. During this second, zenoh continues to operate, but some subscribers in the system may miss some data samples. An end-to-end reliability can be set up to overcome this problem.

Please emphasize this on your website more!  I didn't see any of this on your website, would you be willing to post numbers, plots, etc.?  This is the kind of information that engineers need to see when making decisions between technologies..

**Comment by OlivierHecart** [comment]
The [zenoh.io](https://zenoh.io) clearly needs to be improved on a lot of aspects and dynamic networks support is obviously one of them. 
Note that we tend to publish figures and plots in the [blog](https://zenoh.io/blog/2021-07-05-zenoh-overhead/) section of the site. You can already find some figures on discovery traffic there. We are about to publish a post with plots and figures of zenoh throughput and latency. And we plan to publish something soon about zenoh routing which should include figures on zenoh topology alignment figures.

In the mean time here are some figures from last April from the following scenario:
1. Startup zenoh routers with a given network topology.
2. Measure time until all nodes agree on the topology and computed their routing tables.
3. Kill on of the routers.
4. Measure time until all nodes agree on the new topology and recomputed their routing tables.
5. Restart the killed router.
6. Measure time until all nodes agree on the new topology and recomputed their routing tables.

This has been run on a local host so with very low latency. On a network with more realistic latencies, times will probably increase a bit.
Again, note that during the alignment time, zenoh continues to operate and route data, but as the routes are not up to date some of the subscribers may miss some samples.

![2021 04 07-zenoh-routing-convergence-figures 001](https://user-images.githubusercontent.com/1199006/125062064-4adf1f80-e0ae-11eb-9140-99cd225f5b3b.jpeg)

**Comment by ckaran** [comment]
@OlivierHecart Are you able to selectively break and bring up links between nodes in your simulations?  While nodes sometimes come and go, it's much more common for the node to be operating as normal, but due to environmental constraints (multipath fading, etc.) some, but not all, radio communications links fade out.  Being able to model the communications graph as a fully connected digraph with probabilities of delivery of packets across a given link being controllable over time will give you a better model of actual communications.

Also, I just finally read this [blog post](https://zenoh.io/blog/2021-04-28-ros2-integration/) talking about ROS2 integration, which is the *precise* use-case that I was talking about earlier (I would like to have Zenoh be the communications backbone across a wireless network, assuming that would work).

**Comment by OlivierHecart** [comment]
@ckaran No, we can't do that in our simplistic simulation environment right now. But this is something we'll have to test and characterize as well.
